File: load_sorted_results/kiasort_load_results.py
"""Load kiaSort outputs into Python."""

from __future__ import annotations

import warnings
from pathlib import Path
from typing import Any, Dict, List, Optional

import numpy as np
import h5py
import scipy.io as sio


def kiasort_load_results(
    results_path,
    curated: bool = False,
    individual_waveform: bool = False,
) -> Dict[str, Any]:
    """Load kiaSort outputs.

    Parameters
    ----------
    results_path : str | Path
        The outputFolder passed to the sorter.
    curated : bool
        Prefer RES_Sorted/<field>_curated.h5; fall back to non-curated
        with a warning if missing.
    individual_waveform : bool
        Also load per-spike waveforms when saved.

    Returns
    -------
    dict with keys: source, spike_idx, channelNum, unifiedLabels,
    [labelOnChannel], [waveforms], units (list of dicts with id,
    channel, meanWaveform, isolation).
    """
    results_path = Path(results_path)
    res_folder = results_path / "RES_Sorted"
    if not res_folder.is_dir():
        raise FileNotFoundError(f"RES_Sorted not found at {res_folder}")

    suffix = ""
    if curated:
        if (res_folder / "spike_idx_curated.h5").exists() and \
           (res_folder / "unifiedLabels_curated.h5").exists():
            suffix = "_curated"
        else:
            warnings.warn(
                f"Curated outputs missing in {res_folder}. Loading non-curated."
            )

    out: Dict[str, Any] = {
        "source":        "RES_Sorted" + suffix,
        "spike_idx":     _read_h5(res_folder, f"spike_idx{suffix}"),
        "channelNum":    _read_h5(res_folder, f"channelNum{suffix}"),
        "unifiedLabels": _read_h5(res_folder, f"unifiedLabels{suffix}"),
    }

    labels_file = res_folder / "labels.h5"
    if labels_file.exists():
        out["labelOnChannel"] = _read_h5(res_folder, "labels")

    if individual_waveform:
        wf_file = res_folder / f"waveforms{suffix}.h5"
        if wf_file.exists():
            with h5py.File(wf_file, "r") as f:
                out["waveforms"] = np.asarray(f[f"/waveforms{suffix}"])
        else:
            warnings.warn(f"Waveform file {wf_file} not found.")

    out["units"] = _load_units(results_path, res_folder, suffix)
    return out


def _read_h5(folder: Path, name: str) -> np.ndarray:
    with h5py.File(folder / f"{name}.h5", "r") as f:
        return np.asarray(f[f"/{name}"]).flatten()


def _load_units(results_path: Path, res_folder: Path, suffix: str) -> List[Dict[str, Any]]:
    units: List[Dict[str, Any]] = []

    curated_mat = res_folder / "curated_sample.mat"
    sorted_samp = results_path / "Sorted_Samples" / "sorted_samples.mat"

    if suffix == "_curated" and curated_mat.exists():
        d = sio.loadmat(str(curated_mat), simplify_cells=True)
        c = d.get("curatedSamples")
        if isinstance(c, dict) and "unifiedLabels" in c:
            ids   = np.atleast_1d(c["unifiedLabels"]).flatten()
            chans = np.atleast_1d(c.get("channelNum", np.full(ids.shape, np.nan))).flatten()
            waves = c.get("waveform")
            iso   = c.get("unitIsolation")
            for k, lab in enumerate(ids):
                u: Dict[str, Any] = {"id": _to_py(lab)}
                if k < chans.size:
                    u["channel"] = _to_py(chans[k])
                if isinstance(waves, np.ndarray) and waves.ndim >= 3 and k < waves.shape[0]:
                    u["meanWaveform"] = np.asarray(waves[k])
                if iso is not None:
                    iso_arr = np.atleast_1d(iso).flatten()
                    if k < iso_arr.size:
                        u["isolation"] = str(iso_arr[k])
                units.append(u)
            return units

    if sorted_samp.exists():
        d = sio.loadmat(str(sorted_samp), simplify_cells=True)
        ccs = d.get("crossChannelStats")
        if isinstance(ccs, dict) and "unified_labels" in ccs:
            uL = ccs["unified_labels"]
            ids   = np.atleast_1d(uL["label"]).flatten()
            chans = np.atleast_1d(uL.get("channelID", np.full(ids.shape, np.nan))).flatten()
            waves = uL.get("meanWaveforms")
            for k, lab in enumerate(ids):
                u = {"id": _to_py(lab)}
                if k < chans.size:
                    u["channel"] = _to_py(chans[k])
                if isinstance(waves, np.ndarray) and waves.ndim >= 3 and k < waves.shape[0]:
                    u["meanWaveform"] = np.asarray(waves[k])
                units.append(u)

    return units


def _to_py(x):
    try:
        if np.isnan(x):
            return None
    except (TypeError, ValueError):
        pass
    if hasattr(x, "item"):
        return x.item()
    return x

File: load_sorted_results/kiaSort_load_results.m
function out = kiaSort_load_results(resultsPath, curated, individual_waveform)
% Load kiaSort outputs.
%   resultsPath          - the outputFolder used during sorting
%   curated              - true => prefer RES_Sorted/<field>_curated.h5;
%                          fall back to non-curated with a warning if missing
%   individual_waveform  - true => also load per-spike waveforms when saved

if nargin < 2 || isempty(curated),             curated = false; end
if nargin < 3 || isempty(individual_waveform), individual_waveform = false; end

resFolder = fullfile(resultsPath, 'RES_Sorted');
if ~exist(resFolder, 'dir')
    error('RES_Sorted folder not found at %s', resFolder);
end

suffix = '';
if curated
    if exist(fullfile(resFolder, 'spike_idx_curated.h5'),     'file') && ...
       exist(fullfile(resFolder, 'unifiedLabels_curated.h5'), 'file')
        suffix = '_curated';
    else
        warning('kiaSort_load_results:noCurated', ...
            'Curated outputs missing in %s. Loading non-curated.', resFolder);
    end
end

out = struct();
out.source       = ['RES_Sorted' suffix];
out.spike_idx    = readH5(resFolder, ['spike_idx'     suffix]);
out.channelNum   = readH5(resFolder, ['channelNum'    suffix]);
out.unifiedLabels = readH5(resFolder, ['unifiedLabels' suffix]);

labelsFile = fullfile(resFolder, 'labels.h5');
if exist(labelsFile, 'file')
    out.labelOnChannel = readH5(resFolder, 'labels');
end

if individual_waveform
    wfFile = fullfile(resFolder, ['waveforms' suffix '.h5']);
    if exist(wfFile, 'file')
        out.waveforms = h5read(wfFile, ['/waveforms' suffix]);
    else
        warning('kiaSort_load_results:noWaveforms', ...
            'Waveform file %s not found.', wfFile);
    end
end

out.units = loadUnits(resultsPath, resFolder, suffix);
end


function v = readH5(folder, name)
v = h5read(fullfile(folder, [name '.h5']), ['/' name]);
v = v(:);
end


function units = loadUnits(resultsPath, resFolder, suffix)
units = struct('id', {}, 'channel', {}, 'meanWaveform', {}, 'isolation', {});

curatedMat = fullfile(resFolder, 'curated_sample.mat');
sortedSamp = fullfile(resultsPath, 'Sorted_Samples', 'sorted_samples.mat');

if strcmp(suffix, '_curated') && exist(curatedMat, 'file')
    s = load(curatedMat);
    if isfield(s, 'curatedSamples')
        c = s.curatedSamples;
        ids = c.unifiedLabels(:);
        for k = 1:numel(ids)
            units(k).id      = ids(k);
            if isfield(c, 'channelNum')   && numel(c.channelNum)   >= k
                units(k).channel = c.channelNum(k);
            end
            if isfield(c, 'waveform') && size(c.waveform, 1) >= k
                units(k).meanWaveform = squeeze(c.waveform(k, :, :));
            end
            if isfield(c, 'unitIsolation') && numel(c.unitIsolation) >= k
                units(k).isolation = c.unitIsolation{k};
            end
        end
        return;
    end
end

if exist(sortedSamp, 'file')
    s = load(sortedSamp);
    if isfield(s, 'crossChannelStats') && isfield(s.crossChannelStats, 'unified_labels')
        u = s.crossChannelStats.unified_labels;
        ids = u.label(:);
        for k = 1:numel(ids)
            units(k).id = ids(k);
            if isfield(u, 'channelID')     && numel(u.channelID)     >= k
                units(k).channel = u.channelID(k);
            end
            if isfield(u, 'meanWaveforms') && size(u.meanWaveforms, 1) >= k
                units(k).meanWaveform = squeeze(u.meanWaveforms(k, :, :));
            end
        end
    end
end
end
